### PR TITLE
Sanitize error message for sensitive requests

### PR DESCRIPTION
**Pull request by tustvold**
This avoids credentials potentially leaking into log messages

### CODE REVIEW

File: object_store/src/aws/credential.rs [hunk 1]
```diff
@@ -610,6 +610,7 @@ async fn web_identity(
         ])
         .retryable(retry_config)
         .idempotent(true)
+        .sensitive(true)
```
Comment (alamb) [inline object_store/src/aws/credential.rs:613]: I wonder if similar treatment is needed for ?

https://github.com/apache/arrow-rs/blob/ddec83694d6284fcdfb01b03fec8784edf5038db/object_store/src/aws/credential.rs#L233

https://github.com/apache/arrow-rs/blob/ddec83694d6284fcdfb01b03fec8784edf5038db/object_store/src/aws/credential.rs#L261

**Review by alamb — APPROVED** [review]
Makes sense to me. Thank you @tustvold 

I am not sure if other URLs may need the same treatment. We can always do so in a follow on PR if needed.

File: object_store/src/aws/credential.rs (hunk 1 above)
Comment (tustvold) [reply object_store/src/aws/credential.rs:613]: No neither are credentials and are designed to be exposed to end-users - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html

**Comment by maxburke** [comment]
Is it possible to have this fix applied to the 0.9.x line?

**Comment by alamb** [comment]
> Is it possible to have this fix applied to the 0.9.x line?

The patch should apply cleanly. 

I made a branch https://github.com/apache/arrow-rs/tree/object_store_0.9-dev

If you make a PR targeting that branch I can run the release process to make a 0.9.2 release

**Comment by maxburke** [comment]
Hi Andrew,

Thanks, but I am unable to apply the patch to 0.9 cleanly; it seems to conflict with 0b0d64b6279ac753ec91e32a12b9ee711b6fea29 and 873176078abe54b4503ef15b8d1a2509095331e9.

**Comment by maxburke** [comment]
@alamb / @tustvold is it possible to get a release for this for the 0.9 branch? I don't want to have to spend a week doing a datafusion etc upgrade just to get the fix for this CVE so that our `cargo audit` builds can once again succeed.

Thanks.

**Comment by alamb** [comment]
> @alamb / @tustvold is it possible to get a release for this for the 0.9 branch? I don't want to have to spend a week doing a datafusion etc upgrade just to get the fix for this CVE so that our `cargo audit` builds can once again succeed.

I am willing to run the release process for 0.9.0

However I am not sure I will have time to work on the backport of the code to the 0.9.0 branch. If someone else can do that I will handle the logicistics of releasing

**Comment by alamb** [comment]
@maxburke  -- Note this CVE only applies when using a certain form of authentication with S3. If you aren't using that form the logs won't contain credentials

**Comment by maxburke** [comment]
@alamb We use `cargo audit` to report on vulnerabilities in our dependency chain. It doesn't know enough to know if the afflicted code paths are used by us or not. The result though is all of our Rust builds are on hold because there's a CVE reported against the versions of `object_store` on which we and our dependencies (both `datafusion` and `lancedb`) are using, and because of that our CI pipeline is failing.

The Cargo version resolution rules do not accept 0.10 as being a compatible update for 0.9; I don't think releasing an incompatible version that fixes the issue is acceptable.

I also do not believe it is onerous to ask the package maintainers to release a _compatible patch_ for something serious enough to warrant a CVE number.

**Comment by tustvold** [comment]
`cargo audit` has functionality to suppress advisories if you aren't impacted by them - https://github.com/rustsec/rustsec/blob/main/cargo-audit/README.md#ignoring-advisories

> I also do not believe it is excessively onerous to ask the package maintainers to release a compatible patch for something serious enough to warrant a CVE number.

If you are willing to help backport this fix, Andrew has kindly offered to cut a release for you. We're all volunteers here, I would ask you understand that our capacity and time is limited. 

FWIW a CVE, much like RUSTSEC, is simply an advisory mechanism, in both cases it is expected that downstream projects assess said advisories, determine to what extent they impact them, and act accordingly. In this case a CVE was filed because it was reported via the apache security team, if it had been reported in a different manner it might have been a RUSTSEC advisory instead, there isn't anything special about a CVE that makes it warrant different handling from any other advisory.

**Pull request merged**